### PR TITLE
Reg compacted datasets in Hive

### PR DESCRIPTION
**Pull request by sahilTakiar**
* Initial PR to discuss the approach for adding Hive-Reg for datasets produced by `MRCompactor`
* Registration is done in a `AfterCompactionAction` once compaction has completed
* The `AfterCompactionAction`s are configurable and are invoked in `JobRunnerExecutor.afterExecute`

@zliu41 can you review?

### CODE REVIEW

**Comment by zliu41** [comment]
This creates a HiveMetaStoreClient connection for each dataset. Can you create a HiveRegister object in MRCompactor and use it to register all datasets?

**Comment by sahilTakiar** [comment]
@zliu41 addressed comments and modified approach slightly

* Instead of making a listener specific to `MRCompactor` I made a generic `CompactorListener`
* Added factory interfaces and implementations for creating `Compactor`s and `CompactorListener`s
* Re-factored `AzkabanCompactionJobLauncher` a bit to account for changes
* Main logic for this feature is in the class `HiveRegistrationCompactorListener`

**Comment by zliu41** [comment]
We need a solution for retrying failed Hive registration, to avoid a dataset being published but not registered with Hive. Ideally this should leverage exactly-once, but since it likely won't be ready by Q1, a temp solution is needed, e.g., write failed dataset paths to an HDFS location which can be picked up by the next run to retry.

File: gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java [hunk 1]
```diff
@@ -788,6 +792,13 @@ protected void afterExecute(Runnable r, Throwable t) {
             // Set the dataset status to COMPACTION_COMPLETE if compaction is successful.
             jobRunner.getDataset().setState(COMPACTION_COMPLETE);
           }
+          if (MRCompactor.this.compactorListener.isPresent()) {
```
Comment (zliu41) [inline gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java]: This seems to allow only one listener. Can you make it possible to have multiple listeners?

Comment (sahilTakiar) [inline gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java]: Updated, it can now run multiple listeners serially.

**Comment by sahilTakiar** [comment]
Yes, I will focus on the retry logic in the next PR as it may require some additional design work.

File: gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java (hunk 1 above)
Comment (zliu41) [inline gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java]: What is the change? seems there is still only one listener here. I thought `MRCompactor` should contain a collection of listeners.

Comment (sahilTakiar) [inline gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java]: There is a new class `SerialCompactorListener` that takes in a `List` of `CompactorListener`s and executes them one by one. This class is used by `ReflectionCompactorListenerFactory`.

**Comment by zliu41** [comment]
LGTM, please add copyright block to all new files

**Pull request merged**